### DISCUSSION TITLE
How to detect navigator.platform in Next JS ?

### DISCUSSION THREAD

**Discussion by surjithctly** [vercel/next.js#18197] [Help] [answered]
I'm writing a small function in Next JS with react. When I use navigator, I get the following error.

`ReferenceError: navigator is not defined`

Here's my code:

```jsx
import React from "react";

export default function App() {

const isMac = navigator.platform.toUpperCase().indexOf("MAC") >= 0;

  return (
   {isMac ? "I'm mac" : "I'm windows"}
  );
}
```
I know this renders on server that's why throwing error. 
But how do I make it run on the browser to get the platform and then render a section based on that?

**Comment by jamesmosier** [comment, marked as answer]
Hi @surjithctly. If you only want something to run on the client, you could do:

```js
const isMac = typeof window !== 'undefined' ? navigator.platform.toUpperCase().indexOf("MAC") >= 0 : false;
```

This ensures no browser code runs on the server!

---


If you did not want the entire component to render on the server, you could use `next/dynamic` with `ssr: false`: https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr

**Reply by surjithctly** [reply]
Nice. It works. Thanks

**Reply by PeerRich** [reply]
navigator.platform is deprecated tho, any concerns / alternatives?

**Reply by pitieu** [reply]
You can now do it this way

```
const isMac =
    typeof window !== "undefined"
      ? navigator.userAgent.toUpperCase().indexOf("MAC") >= 0
      : false;
```

**Reply by jonasteich** [reply]
@pitieu When I do it like this, I get an Hydration failed error because the initial UI does not match what was rendered on the server. What is the problem?

**Reply by JasonLamv-t** [reply]
@JonasTeich I think it's because the `window` was `undefined` in the server. 
Just try `const isMac = /(Mac|iPhone|iPod|iPad)/i.test(navigator.userAgent);`, if it works, let it go.

**Reply by hc0503** [reply]
@jamesmosier ,
Can we detect Windows OS? and Mobile, Android, IOS, Browser type?
